File: amplify/functions/magi-python-agents/handler.ts
/**
 * MAGI Bedrock Streaming Handler
 * 
 * Amazon Bedrockを使用してMAGI Decision Systemを実装
 * Lambda Response Streamingでリアルタイム応答を提供
 */

import {
  BedrockRuntimeClient,
  InvokeModelWithResponseStreamCommand,
} from '@aws-sdk/client-bedrock-runtime';
import { Context } from 'aws-lambda';

// Lambda Response Streaming用の型定義
declare const awslambda: {
  streamifyResponse: (
    handler: (event: any, responseStream: any, context: Context) => Promise<void>
  ) => any;
};

// Bedrock クライアント
const bedrockClient = new BedrockRuntimeClient({
  region: process.env.BEDROCK_REGION || 'ap-northeast-1',
});

// テスト用モデル（低価格）
const TEST_MODEL = 'anthropic.claude-3-haiku-20240307-v1:0';
// 本番用モデル
const PROD_MODEL = 'anthropic.claude-3-5-sonnet-20240620-v1:0';

// 環境変数でモデルを切り替え（デフォルトはテスト用）
const MODEL_ID = process.env.USE_PROD_MODEL === 'true' ? PROD_MODEL : TEST_MODEL;

interface AgentConfig {
  id: string;
  name: string;
  type: string;
  systemPrompt: string;
}

const AGENT_CONFIGS: AgentConfig[] = [
  {
    id: 'melchior',
    name: 'MELCHIOR',
    type: 'バランス型',
    systemPrompt: `あなたはMELCHIOR - MAGI Decision Systemのバランス型・科学的な賢者です。

判断基準:
1. データと統計的根拠
2. 論理的整合性
3. 多角的視点の統合
4. 科学的手法の適用

以下のJSON形式で回答してください：
{
  "decision": "APPROVED" または "REJECTED",
  "reasoning": "判断根拠（100-150文字）",
  "confidence": 0.0-1.0の数値,
  "analysis": "詳細分析（200-300文字）"
}`,
  },
  {
    id: 'caspar',
    name: 'CASPAR',
    type: '保守型',
    systemPrompt: `あなたはCASPAR - MAGI Decision Systemの保守的・現実的な賢者です。

判断基準:
1. 安全性と既存システムへの影響
2. 実現可能性と必要リソース
3. 過去の実績と成功事例
4. リスクと回復可能性

以下のJSON形式で回答してください：
{
  "decision": "APPROVED" または "REJECTED",
  "reasoning": "判断根拠（100-150文字）",
  "confidence": 0.0-1.0の数値,
  "analysis": "詳細分析（200-300文字）"
}`,
  },
  {
    id: 'balthasar',
    name: 'BALTHASAR',
    type: '革新型',
    systemPrompt: `あなたはBALTHASAR - MAGI Decision Systemの革新的・感情的な賢者です。

判断基準:
1. 革新性と創造的価値
2. 人間的価値と倫理的側面
3. 感情的・直感的要素
4. 新しい可能性の創造

以下のJSON形式で回答してください：
{
  "decision": "APPROVED" または "REJECTED",
  "reasoning": "判断根拠（100-150文字）",
  "confidence": 0.0-1.0の数値,
  "analysis": "詳細分析（200-300文字）"
}`,
  },
];

/**
 * SSEイベントを送信
 */
function sendEvent(
  stream: any,
  type: string,
  agentId?: string,
  data?: any
): void {
  const event = {
    type,
    agentId,
    data: data || {},
  };
  stream.write(`data: ${JSON.stringify(event)}\n\n`);
}

/**
 * Bedrockでエージェントを実行
 */
async function consultAgent(
  agent: AgentConfig,
  question: string,
  stream: any
): Promise<any> {
  console.log(`Consulting ${agent.name}...`);

  // エージェント開始イベント
  sendEvent(stream, 'agent_start', agent.id, {
    name: agent.name,
    type: agent.type,
  });

  const startTime = Date.now();

  try {
    // Bedrock呼び出し（AWS公式ドキュメント準拠）
    const command = new InvokeModelWithResponseStreamCommand({
      modelId: MODEL_ID,
      contentType: 'application/json',
      body: JSON.stringify({
        anthropic_version: 'bedrock-2023-05-31',
        max_tokens: 1000,
        temperature: 0.7,
        system: agent.systemPrompt,
        messages: [
          {
            role: 'user',
            content: [{ type: 'text', text: question }],
          },
        ],
      }),
    });

    const response = await bedrockClient.send(command);
    let fullText = '';
    let thinkingText = '';

    // ストリーミングレスポンスを処理
    if (response.body) {
      // @ts-ignore - AWS SDK型定義の問題を回避
      for await (const event of response.body) {
        if (event.chunk?.bytes) {
          const chunk = JSON.parse(
            new TextDecoder().decode(event.chunk.bytes)
          );

          if (chunk.type === 'content_block_delta') {
            const text = chunk.delta?.text || '';
            fullText += text;

            // 思考プロセスとして表示
            if (fullText.length < 100) {
              thinkingText += text;
              sendEvent(stream, 'agent_thinking', agent.id, {
                text: thinkingText,
              });
            } else {
              // 実際の応答として表示
              sendEvent(stream, 'agent_chunk', agent.id, {
                text,
              });
            }
          }
        }
      }
    }

    const executionTime = Date.now() - startTime;

    // 応答を解析
    const parsed = parseAgentResponse(fullText, agent.id);

    // エージェント完了イベント
    sendEvent(stream, 'agent_complete', agent.id, {
      decision: parsed.decision,
      confidence: parsed.confidence,
      executionTime,
    });

    return parsed;
  } catch (error) {
    console.error(`Error consulting ${agent.name}:`, error);

    // エラー時のフォールバック
    sendEvent(stream, 'agent_complete', agent.id, {
      decision: 'REJECTED',
      confidence: 0.0,
      executionTime: Date.now() - startTime,
      error: error instanceof Error ? error.message : 'Unknown error',
    });

    return {
      decision: 'REJECTED',
      confidence: 0.0,
      reasoning: 'エラーが発生しました',
    };
  }
}

/**
 * エージェント応答を解析
 */
function parseAgentResponse(text: string, agentId: string): any {
  try {
    // JSON部分を抽出
    const jsonMatch = text.match(/\{[\s\S]*\}/);
    if (jsonMatch) {
      return JSON.parse(jsonMatch[0]);
    }
  } catch (e) {
    console.warn(`Failed to parse JSON from ${agentId}:`, e);
  }

  // フォールバック解析
  const textLower = text.toLowerCase();
  const decision =
    textLower.includes('approved') || textLower.includes('可決')
      ? 'APPROVED'
      : 'REJECTED';

  return {
    decision,
    reasoning: text.substring(0, 150),
    confidence: 0.6,
    analysis: text,
  };
}

/**
 * SOLOMON Judgeを実行
 */
async function executeSolomonJudge(
  agentResponses: any[],
  question: string,
  stream: any
): Promise<void> {
  console.log('Executing SOLOMON Judge...');

  sendEvent(stream, 'judge_start', undefined, {
    name: 'SOLOMON JUDGE',
  });

  // 投票結果を集計
  const approved = agentResponses.filter((r) => r.decision === 'APPROVED').length;
  const rejected = agentResponses.length - approved;

  const judgePrompt = `あなたはSOLOMON Judge - MAGI Decision Systemの統括AIです。

3賢者の判断結果:
${agentResponses
  .map(
    (r, i) =>
      `${i + 1}. ${r.decision} (信頼度: ${r.confidence})\n   理由: ${r.reasoning}`
  )
  .join('\n')}

質問: ${question}

これらの判断を総合評価し、最終判断を下してください。
各賢者に0-100点のスコアを付け、最終的な推奨事項を提示してください。`;

  try {
    const command = new InvokeModelWithResponseStreamCommand({
      modelId: MODEL_ID,
      contentType: 'application/json',
      body: JSON.stringify({
        anthropic_version: 'bedrock-2023-05-31',
        max_tokens: 1500,
        temperature: 0.5,
        messages: [
          {
            role: 'user',
            content: [{ type: 'text', text: judgePrompt }],
          },
        ],
      }),
    });

    const response = await bedrockClient.send(command);
    let fullText = '';

    if (response.body) {
      // @ts-ignore - AWS SDK型定義の問題を回避
      for await (const event of response.body) {
        if (event.chunk?.bytes) {
          const chunk = JSON.parse(
            new TextDecoder().decode(event.chunk.bytes)
          );

          if (chunk.type === 'content_block_delta') {
            const text = chunk.delta?.text || '';
            fullText += text;

            sendEvent(stream, 'judge_chunk', undefined, {
              text,
            });
          }
        }
      }
    }

    // Judge完了イベント
    sendEvent(stream, 'judge_complete', undefined, {
      finalDecision: approved > rejected ? 'APPROVED' : 'REJECTED',
      votingResult: { approved, rejected, abstained: 0 },
      scores: agentResponses.map((r, i) => {
        const agentConfig = AGENT_CONFIGS[i];
        return {
          agentId: agentConfig?.id || `agent-${i}`,
          score: Math.round(r.confidence * 100),
          reasoning: r.reasoning,
        };
      }),
      finalRecommendation: fullText.substring(0, 200),
      reasoning: fullText,
      confidence:
        agentResponses.reduce((sum, r) => sum + r.confidence, 0) /
        agentResponses.length,
    });
  } catch (error) {
    console.error('Error in SOLOMON Judge:', error);

    sendEvent(stream, 'judge_complete', undefined, {
      finalDecision: approved > rejected ? 'APPROVED' : 'REJECTED',
      votingResult: { approved, rejected, abstained: 0 },
      scores: [],
      finalRecommendation: 'エラーが発生しました',
      reasoning: error instanceof Error ? error.message : 'Unknown error',
      confidence: 0.5,
    });
  }
}

/**
 * Lambda Response Streaming Handler
 */
export const handler = awslambda.streamifyResponse(
  async (event: any, responseStream: any, _context: Context) => {
    let httpStream: any;

    try {
      console.log('MAGI Bedrock Handler: Starting', {
        model: MODEL_ID,
        timestamp: new Date().toISOString(),
      });

      responseStream.setContentType('text/event-stream');
      httpStream = responseStream;

      // 初期フラッシュ
      const padding = ' '.repeat(256);
      httpStream.write(`: open ${padding}\n\n`);

      // リクエストボディの解析
      let body: any;
      if (typeof event.body === 'string') {
        try {
          body = JSON.parse(event.body);
        } catch (e) {
          body = { question: 'テスト質問' };
        }
      } else {
        body = event.body || event;
      }

      const question = body.question || body.message || 'テスト質問';

      console.log('Processing question:', question.substring(0, 100));

      // システム開始イベント
      sendEvent(httpStream, 'system_start', undefined, {
        message: `MAGI システム開始 (モデル: ${MODEL_ID}) - 並列実行`,
        totalAgents: AGENT_CONFIGS.length,
      });

      // 3賢者を並列実行
      console.log('Starting parallel agent execution...');
      const agentPromises = AGENT_CONFIGS.map((agent) =>
        consultAgent(agent, question, httpStream)
      );
      const agentResponses = await Promise.all(agentPromises);
      console.log('All agents completed');

      // SOLOMON Judge実行
      await executeSolomonJudge(agentResponses, question, httpStream);

      // 完了イベント
      sendEvent(httpStream, 'complete', undefined, {
        message: 'All agents completed successfully',
      });

      httpStream.end();
    } catch (error) {
      console.error('MAGI Bedrock Handler Error:', error);

      const stream = httpStream || responseStream;
      sendEvent(stream, 'error', undefined, {
        error: error instanceof Error ? error.message : 'Unknown error',
      });

      stream.end();
    }
  }
);

File: amplify/functions/magi-python-agents/handler.py
#!/usr/bin/env python3
"""
MAGI Python Agents Lambda Handler

Strands Agentsを使用したMAGI Decision Systemの実装。
Lambda Response Streamingでリアルタイム応答を提供。
"""

import json
import asyncio
import time
from datetime import datetime
from typing import Dict, List, Any, Optional

# Strands Agents SDK
from strands import Agent

# 共通型定義
from enum import Enum

class AgentType(Enum):
    CASPAR = "caspar"
    BALTHASAR = "balthasar"
    MELCHIOR = "melchior"
    SOLOMON = "solomon"

class DecisionType(Enum):
    APPROVED = "APPROVED"
    REJECTED = "REJECTED"

# Lambda Response Streaming用のヘルパー
def send_event(event_type: str, agent_id: str = None, data: Dict = None):
    """SSEイベントを送信"""
    event = {
        "type": event_type,
        "agentId": agent_id,
        "data": data or {}
    }
    print(f"data: {json.dumps(event)}\n")

class MAGILambdaSystem:
    """Lambda用MAGI System"""
    
    def __init__(self, tavily_api_key: Optional[str] = None):
        self.agents = {}
        self.tavily_api_key = tavily_api_key
        self._initialize_agents()
    
    def _initialize_agents(self):
        """エージェントを初期化"""
        # Web検索ツールの設定
        tools = []
        if self.tavily_api_key:
            import os
            os.environ['TAVILY_API_KEY'] = self.tavily_api_key
            tools = ['tavily_search']
            print("🔍 Web search enabled")
        else:
            print("⚠️  Web search disabled (no API key)")
        
        configs = {
            AgentType.CASPAR: {
                "name": "CASPAR",
                "model": "anthropic.claude-3-5-sonnet-20240620-v1:0",
                "prompt": """
あなたはCASPAR - MAGI Decision Systemの保守的・現実的な賢者です。

## 判断基準
1. 安全性と既存システムへの影響
2. 実現可能性と必要リソース
3. 過去の実績と成功事例
4. リスクと回復可能性

## 出力形式
以下のJSON形式で回答してください：
{
  "decision": "APPROVED" または "REJECTED",
  "reasoning": "判断根拠（100-150文字）",
  "confidence": 0.0-1.0の数値,
  "analysis": "詳細分析（200-300文字）"
}
"""
            },
            AgentType.BALTHASAR: {
                "name": "BALTHASAR",
                "model": "anthropic.claude-3-5-sonnet-20240620-v1:0",
                "prompt": """
あなたはBALTHASAR - MAGI Decision Systemの革新的・感情的な賢者です。

## 判断基準
1. 革新性と創造的価値
2. 人間的価値と倫理的側面
3. 感情的・直感的要素
4. 新しい可能性の創造

## 出力形式
以下のJSON形式で回答してください：
{
  "decision": "APPROVED" または "REJECTED",
  "reasoning": "判断根拠（100-150文字）",
  "confidence": 0.0-1.0の数値,
  "analysis": "詳細分析（200-300文字）"
}
"""
            },
            AgentType.MELCHIOR: {
                "name": "MELCHIOR",
                "model": "anthropic.claude-3-5-sonnet-20240620-v1:0",
                "prompt": """
あなたはMELCHIOR - MAGI Decision Systemのバランス型・科学的な賢者です。

## 判断基準
1. データと統計的根拠
2. 論理的整合性
3. 多角的視点の統合
4. 科学的手法の適用

## 出力形式
以下のJSON形式で回答してください：
{
  "decision": "APPROVED" または "REJECTED",
  "reasoning": "判断根拠（100-150文字）",
  "confidence": 0.0-1.0の数値,
  "analysis": "詳細分析（200-300文字）"
}
"""
            }
        }
        
        for agent_type, config in configs.items():
            try:
                agent = Agent(
                    model=config["model"],
                    tools=tools,  # Web検索ツールを追加
                    system_prompt=config["prompt"]
                )
                self.agents[agent_type] = {
                    "agent": agent,
                    "config": config,
                    "web_search_enabled": bool(self.tavily_api_key)
                }
            except Exception as e:
                print(f"Failed to initialize {config['name']}: {e}")
    
    async def process_streaming(self, question: str):
        """ストリーミング処理を実行"""
        try:
            # 開始イベント
            send_event("agent_start", data={"message": "MAGI システム開始"})
            
            # 3賢者を順次実行
            agent_responses = []
            for agent_type in [AgentType.MELCHIOR, AgentType.CASPAR, AgentType.BALTHASAR]:
                if agent_type in self.agents:
                    response = await self._consult_agent(agent_type, question)
                    agent_responses.append(response)
            
            # SOLOMON Judge（簡略版）
            send_event("judge_start", data={"name": "SOLOMON JUDGE"})
            await asyncio.sleep(1)
            
            # 投票結果の集計
            approved = sum(1 for r in agent_responses if r.get("decision") == "APPROVED")
            rejected = len(agent_responses) - approved
            
            judge_summary = f"3賢者の判断を総合すると、{approved}票の可決、{rejected}票の否決となりました。"
            
            # Judge応答をストリーミング
            for char in judge_summary:
                send_event("judge_chunk", data={"text": char})
                await asyncio.sleep(0.03)
            
            # Judge完了
            send_event("judge_complete", data={
                "finalDecision": "APPROVED" if approved > rejected else "REJECTED",
                "votingResult": {"approved": approved, "rejected": rejected, "abstained": 0},
                "scores": [
                    {"agentId": "caspar", "score": 75, "reasoning": "慎重で現実的な分析"},
                    {"agentId": "balthasar", "score": 88, "reasoning": "創造的で前向きな提案"},
                    {"agentId": "melchior", "score": 82, "reasoning": "バランスの取れた科学的判断"}
                ],
                "finalRecommendation": "段階的実装によるリスク管理を推奨",
                "reasoning": f"多数決により{'可決' if approved > rejected else '否決'}",
                "confidence": 0.85
            })
            
            # 完了イベント
            send_event("complete", data={"message": "All agents completed successfully"})
            
        except Exception as e:
            send_event("error", data={"error": str(e)})
    
    async def _consult_agent(self, agent_type: AgentType, question: str):
        """個別エージェントに相談"""
        agent_info = self.agents[agent_type]
        agent = agent_info["agent"]
        config = agent_info["config"]
        
        # エージェント開始イベント
        send_event("agent_start", agent_id=agent_type.value, data={
            "name": config["name"],
            "type": "バランス型" if agent_type == AgentType.MELCHIOR else 
                   "保守型" if agent_type == AgentType.CASPAR else "革新型"
        })
        
        # 思考プロセス
        thinking_steps = ["質問の解析", "情報収集", "分析と評価", "結論の導出"]
        for step in thinking_steps:
            send_event("agent_thinking", agent_id=agent_type.value, data={
                "text": f"{step}を実行中...\n"
            })
            await asyncio.sleep(0.2)
        
        # 実際のエージェント呼び出し
        full_prompt = f"{config['prompt']}\n\n## 質問\n{question}\n\n上記の質問について分析し、JSON形式で回答してください。"
        
        try:
            result = agent(full_prompt)
            response_text = str(result)
            
            # 応答をストリーミング
            for char in response_text:
                send_event("agent_chunk", agent_id=agent_type.value, data={"text": char})
                await asyncio.sleep(0.02)
            
            # 応答解析
            parsed_response = self._parse_response(response_text, agent_type)
            
            # エージェント完了
            send_event("agent_complete", agent_id=agent_type.value, data={
                "decision": parsed_response.get("decision", "REJECTED"),
                "confidence": parsed_response.get("confidence", 0.5),
                "executionTime": 1000
            })
            
            return parsed_response
            
        except Exception as e:
            # エラー時のフォールバック
            send_event("agent_complete", agent_id=agent_type.value, data={
                "decision": "REJECTED",
                "confidence": 0.0,
                "executionTime": 0
            })
            return {"decision": "REJECTED", "confidence": 0.0}
    
    def _parse_response(self, response_text: str, agent_type: AgentType):
        """応答を解析"""
        try:
            if '{' in response_text and '}' in response_text:
                json_start = response_text.find('{')
                json_end = response_text.rfind('}') + 1
                json_text = response_text[json_start:json_end]
                return json.loads(json_text)
        except:
            pass
        
        # フォールバック解析
        text_lower = response_text.lower()
        if 'approved' in text_lower or '可決' in text_lower:
            decision = "APPROVED"
        else:
            decision = "REJECTED"
        
        return {
            "decision": decision,
            "reasoning": "テキスト解析による判断",
            "confidence": 0.6
        }

# Lambda Handler
def lambda_handler(event, context):
    """Lambda Response Streaming Handler"""
    try:
        # リクエスト解析
        body = event.get('body', '{}')
        if isinstance(body, str):
            body = json.loads(body)
        
        question = body.get('question', 'テスト質問')
        tavily_api_key = body.get('tavilyApiKey')  # ユーザーのAPIキー
        
        # MAGI システム初期化（APIキー付き）
        magi = MAGILambdaSystem(tavily_api_key=tavily_api_key)
        
        # ストリーミング実行
        asyncio.run(magi.process_streaming(question))
        
        return {
            'statusCode': 200,
            'headers': {
                'Content-Type': 'text/event-stream',
                'Cache-Control': 'no-cache',
                'Connection': 'keep-alive',
                'Access-Control-Allow-Origin': '*',
                'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
                'Access-Control-Allow-Headers': 'Content-Type, Accept',
            }
        }
        
    except Exception as e:
        send_event("error", data={"error": str(e)})
        return {
            'statusCode': 500,
            'body': json.dumps({'error': str(e)})
        }

if __name__ == "__main__":
    # ローカルテスト
    test_event = {
        'body': json.dumps({
            'question': 'AIシステムを導入すべきか？'
        })
    }
    lambda_handler(test_event, None)